lisa/schema.py

```diff
@@ -659,35 +683,36 @@ class Environment:
 
     def __post_init__(self, *args: Any, **kwargs: Any) -> None:
         self.nodes: Optional[List[Union[LocalNode, RemoteNode]]] = None
-        if self.nodes_raw is not None:
-            self.nodes = []
-            for node_raw in self.nodes_raw:
-                node_type = node_raw[constants.TYPE]
-                if node_type == constants.ENVIRONMENTS_NODES_LOCAL:
-                    node: Union[
-                        LocalNode, RemoteNode
-                    ] = LocalNode.schema().load(  # type:ignore
-                        node_raw
-                    )
-                    if self.nodes is None:
-                        self.nodes = []
-                    self.nodes.append(node)
-                elif node_type == constants.ENVIRONMENTS_NODES_REMOTE:
-                    node = RemoteNode.schema().load(node_raw)  # type:ignore
-                    if self.nodes is None:
-                        self.nodes = []
-                    self.nodes.append(node)
-                elif node_type == constants.ENVIRONMENTS_NODES_REQUIREMENT:
-                    original_req: NodeSpace = NodeSpace.schema().load(  # type:ignore
-                        node_raw
-                    )
-                    expanded_req = original_req.expand_by_node_count()
-                    if self.nodes_requirement is None:
-                        self.nodes_requirement = []
-                    self.nodes_requirement.extend(expanded_req)
+        if not self.nodes_raw:
+            return
+        nodes_final: List[Union[LocalNode, RemoteNode]] = []
+        for node in self.nodes_raw:
+            node_type = node[constants.TYPE]
+            new_node: Union[LocalNode, RemoteNode]
+            if node_type == constants.ENVIRONMENTS_NODES_REQUIREMENT:
+                original_req: NodeSpace = NodeSpace.schema().load(  # type:ignore
+                    node
+                )
+                expanded_req = original_req.expand_by_node_count()
+                if self.nodes_requirement is None:
+                    self.nodes_requirement = []
+                self.nodes_requirement.extend(expanded_req)
+            else:
+                # here we honor LocalNode, RemoteNode and all their extensions
+                subs: Dict[str, Type[Node]] = {}
+                for sub in self._get_subclasses(Node):
+                    subs[sub.type] = sub
+                if node_type not in subs:
+                    raise LisaException(f"unknown node type '{node_type}': {node}")
                 else:
-                    raise LisaException(f"unknown node type '{node_type}': {node_raw}")
-            self.nodes_raw = None
+                    new_node = subs[node_type].schema().load(node)  # type: ignore
+                    nodes_final.append(new_node)
+        self.nodes = nodes_final
+
+    def _get_subclasses(self, t: Type[Node]) -> Iterable[Type[Node]]:
```

Comment:
The initialization of environment and node can be move from the schema to their implementations. That helps the logic is more clear, and the node can be used standalone. Refer to Notifier.

Reply:
I will make a prototype of this.

Reply:
This seems unrelated to my patch

Reply:
It produces duplicate code like this methods. Fixed in my PR.